File: app/src/main/java/com/bandwidth/android/MainActivity.java
package com.bandwidth.android;

import android.Manifest;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.bandwidth.android.app.Conference;
import com.bandwidth.webrtc.RTCBandwidth;
import com.bandwidth.webrtc.RTCBandwidthClient;
import com.bandwidth.webrtc.signaling.ConnectionException;

import org.webrtc.Camera1Enumerator;
import org.webrtc.Camera2Enumerator;
import org.webrtc.CameraEnumerator;
import org.webrtc.EglBase;
import org.webrtc.RendererCommon;
import org.webrtc.SurfaceTextureHelper;
import org.webrtc.SurfaceViewRenderer;
import org.webrtc.VideoCapturer;
import org.webrtc.VideoSource;
import org.webrtc.VideoTrack;

import java.io.IOException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class MainActivity extends AppCompatActivity {
    private SurfaceViewRenderer localRenderer;
    private SurfaceViewRenderer remoteRenderer;

    private RTCBandwidth bandwidth;

    VideoCapturer videoCapturer;

    private VideoTrack localVideoTrack;
    private VideoTrack remoteVideoTrack;
    private HashMap<String, VideoTrack> offeredVideoTracks = new HashMap<String, VideoTrack>();

    private EglBase eglBase;

    private Boolean isConnected = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.CAMERA,Manifest.permission.RECORD_AUDIO}, 100);

        localRenderer = findViewById(R.id.localSurfaceViewRenderer);
        remoteRenderer = findViewById(R.id.remoteSurfaceViewRenderer);

        eglBase = EglBase.create();

        // Create local video renderer.
        localRenderer.init(eglBase.getEglBaseContext(), null);
        localRenderer.setMirror(true);
        localRenderer.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT);

        // Create remote video renderer.
        remoteRenderer.init(eglBase.getEglBaseContext(), null);
        remoteRenderer.setMirror(false);
        remoteRenderer.setEnableHardwareScaler(false);

        bandwidth = new RTCBandwidthClient(getApplicationContext(), eglBase.getEglBaseContext());

        bandwidth.setOnStreamAvailableListener((streamId, mediaTypes, audioTracks, videoTracks, alias) -> {
            if (!videoTracks.isEmpty()) {
                if (!offeredVideoTracks.containsKey(streamId)) {
                    offeredVideoTracks.put(streamId,videoTracks.get(0));
                }
                if (remoteVideoTrack == null) {
                    runOnUiThread(() -> {
                        remoteVideoTrack = videoTracks.get(0);
                        remoteVideoTrack.setEnabled(true);
                        remoteVideoTrack.addSink(remoteRenderer);
                    });
                };
            };
        });

        bandwidth.setOnStreamUnavailableListener(streamId -> {
            offeredVideoTracks.remove(streamId);
            if (offeredVideoTracks.isEmpty()) {
                System.out.println("on Stream NOT available - no more tracks");
                remoteRenderer.clearImage();
                remoteVideoTrack = null;
            } else {
                VideoTrack temp = offeredVideoTracks.entrySet().iterator().next().getValue();
                System.out.println("on Stream NOT available - video track replacement - " +  temp);
                runOnUiThread(() -> {
                    remoteVideoTrack = temp;
                    remoteVideoTrack.setEnabled(true);
                    remoteVideoTrack.addSink(remoteRenderer);
                });
            };
        });

        final Button button = findViewById(R.id.connectButton);
        button.setOnClickListener(view -> {
            if (isConnected) {
                disconnect();
                button.setText("Connect");
            } else {
                connect();
                button.setText("Disconnect");
            }
        });
    }

    private void connect() {
        offeredVideoTracks = new HashMap<String, VideoTrack>();
        new Thread((() -> {
            try {
                String deviceToken = Conference.getInstance().requestDeviceToken(BuildConfig.CONFERENCE_SERVER_PATH);
                bandwidth.connect(BuildConfig.WEBRTC_SERVER_PATH, deviceToken, () -> {
                    isConnected = true;

                    bandwidth.publish("hello-world-android", (streamId, mediaTypes, audioSource, audioTrack, videoSource, videoTrack) -> {
                        runOnUiThread(() -> publish(videoSource, videoTrack));
                    });
                });
            } catch (IOException | ConnectionException e) {
                e.printStackTrace();
            }
        })).start();
    }

    private void disconnect() {
        isConnected = false;

        try {
            videoCapturer.stopCapture();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        localRenderer.clearImage();
        remoteVideoTrack = null;

        bandwidth.disconnect();

        remoteRenderer.clearImage();
    }

    private VideoCapturer createVideoCapturer() {
        CameraEnumerator cameraEnumerator = createCameraEnumerator();

        String[] deviceNames = cameraEnumerator.getDeviceNames();

        for (String deviceName : deviceNames) {
            if (cameraEnumerator.isFrontFacing(deviceName)) {
                return cameraEnumerator.createCapturer(deviceName, null);
            }
        }

        return null;
    }

    private CameraEnumerator createCameraEnumerator() {
        if (Camera2Enumerator.isSupported(getApplicationContext())) {
            return new Camera2Enumerator(getApplicationContext());
        }

        return new Camera1Enumerator(false);
    }

    private void publish(VideoSource videoSource, VideoTrack videoTrack) {
        localVideoTrack = videoTrack;

        SurfaceTextureHelper surfaceTextureHelper = SurfaceTextureHelper.create("CaptureThread", eglBase.getEglBaseContext());

        videoCapturer = createVideoCapturer();
        videoCapturer.initialize(surfaceTextureHelper, getApplicationContext(), videoSource.getCapturerObserver());
        videoCapturer.startCapture(640, 480, 20);

        System.out.println("video stream attributes" + videoTrack);

        localVideoTrack.setEnabled(true);
        localVideoTrack.addSink(localRenderer);
    }

}

File: app/src/main/java/com/bandwidth/android/app/MainActivity.kt
package com.bandwidth.android.app

import android.Manifest
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.bandwidth.android.BuildConfig
import com.bandwidth.android.R
import com.bandwidth.webrtc.RTCBandwidth
import com.bandwidth.webrtc.RTCBandwidthClient
import com.bandwidth.webrtc.listeners.OnConnectListener
import com.bandwidth.webrtc.listeners.OnPublishListener
import com.bandwidth.webrtc.listeners.OnStreamAvailableListener
import com.bandwidth.webrtc.listeners.OnStreamUnavailableListener
import com.bandwidth.webrtc.signaling.ConnectionException
import org.webrtc.*
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private var localRenderer: SurfaceViewRenderer? = null
    private var remoteRenderer: SurfaceViewRenderer? = null
    private var bandwidth: RTCBandwidth? = null
    var videoCapturer: VideoCapturer? = null
    private var localVideoTrack: VideoTrack? = null
    private var remoteVideoTrack: VideoTrack? = null
    private var offeredVideoTracks = HashMap<String, VideoTrack>()
    private var eglBase: EglBase? = null
    private var isConnected = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        this.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        ActivityCompat.requestPermissions(
            this,
            arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO),
            100
        )
        localRenderer = findViewById(R.id.localSurfaceViewRenderer)
        remoteRenderer = findViewById(R.id.remoteSurfaceViewRenderer)
        eglBase = EglBase.create()

        // Create local video renderer.
        localRenderer?.init(eglBase?.getEglBaseContext(), null)
        localRenderer?.setMirror(true)
        localRenderer?.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT)

        // Create remote video renderer.
        remoteRenderer?.init(eglBase?.getEglBaseContext(), null)
        remoteRenderer?.setMirror(false)
        remoteRenderer?.setEnableHardwareScaler(false)
        bandwidth = RTCBandwidthClient(applicationContext, eglBase?.getEglBaseContext())
        bandwidth?.setOnStreamAvailableListener(object : OnStreamAvailableListener {
            override fun onStreamAvailable(
                streamId: String?,
                mediaTypes: List<String?>?,
                audioTracks: List<AudioTrack?>?,
                videoTracks: List<VideoTrack?>?,
                alias: String?
            ) {
                if (videoTracks.isNullOrEmpty().not()) {
                    if (!offeredVideoTracks.containsKey(streamId)) {
                        offeredVideoTracks[streamId] = videoTracks?.get(0)
                    }
                    if (remoteVideoTrack == null) {
                        runOnUiThread {
                            remoteVideoTrack = videoTracks?.get(0)
                            remoteVideoTrack!!.setEnabled(true)
                            remoteVideoTrack!!.addSink(remoteRenderer)
                        }
                    }
                }
            }
        })
        bandwidth?.setOnStreamUnavailableListener(object : OnStreamUnavailableListener {
            override fun onStreamUnavailable(streamId: String?) {
                offeredVideoTracks.remove(streamId)
                if (offeredVideoTracks.isEmpty()) {
                    println("on Stream NOT available - no more tracks")
                    remoteRenderer?.clearImage()
                    remoteVideoTrack = null
                } else {
                    val temp = offeredVideoTracks.entries.iterator().next().value
                    println("on Stream NOT available - video track replacement - $temp")
                    runOnUiThread {
                        remoteVideoTrack = temp
                        remoteVideoTrack!!.setEnabled(true)
                        remoteVideoTrack!!.addSink(remoteRenderer)
                    }
                }
            }
        })
        val button = findViewById<Button>(R.id.connectButton)
        button.setOnClickListener { view: View? ->
            if (isConnected) {
                disconnect()
                button.text = "Connect"
            } else {
                connect()
                button.text = "Disconnect"
            }
        }
    }

    private fun connect() {
        offeredVideoTracks = HashMap()
        Thread {
            try {
                val deviceToken: String? =
                    Conference?.instance?.requestDeviceToken(BuildConfig.CONFERENCE_SERVER_PATH)
                bandwidth!!.connect(BuildConfig.WEBRTC_SERVER_PATH, deviceToken,
                    object : OnConnectListener {
                        override fun onConnect() {
                            isConnected = true
                            bandwidth!!.publish(
                                "hello-world-android",
                                object : OnPublishListener {
                                    override fun onPublish(
                                        streamId: String?,
                                        mediaTypes: List<String?>?,
                                        audioSource: AudioSource?,
                                        audioTrack: AudioTrack?,
                                        videoSource: VideoSource?,
                                        videoTrack: VideoTrack?
                                    ) {
                                        runOnUiThread {
                                            if (videoSource != null) {
                                                if (videoTrack != null) {
                                                    publish(
                                                        videoSource,
                                                        videoTrack
                                                    )
                                                }
                                            }
                                        }
                                    }
                                })
                        }
                    })
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: ConnectionException) {
                e.printStackTrace()
            }
        }.start()
    }

    private fun disconnect() {
        isConnected = false
        try {
            videoCapturer!!.stopCapture()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        localRenderer!!.clearImage()
        remoteVideoTrack = null
        bandwidth!!.disconnect()
        remoteRenderer!!.clearImage()
    }

    private fun createVideoCapturer(): VideoCapturer? {
        val cameraEnumerator = createCameraEnumerator()
        val deviceNames = cameraEnumerator.deviceNames
        for (deviceName in deviceNames) {
            if (cameraEnumerator.isFrontFacing(deviceName)) {
                return cameraEnumerator.createCapturer(deviceName, null)
            }
        }
        return null
    }

    private fun createCameraEnumerator(): CameraEnumerator {
        return if (Camera2Enumerator.isSupported(applicationContext)) {
            Camera2Enumerator(applicationContext)
        } else Camera1Enumerator(false)
    }

    private fun publish(videoSource: VideoSource, videoTrack: VideoTrack) {
        localVideoTrack = videoTrack
        val surfaceTextureHelper =
            SurfaceTextureHelper.create("CaptureThread", eglBase!!.eglBaseContext)
        videoCapturer = createVideoCapturer()
        videoCapturer!!.initialize(
            surfaceTextureHelper,
            applicationContext,
            videoSource.capturerObserver
        )
        videoCapturer!!.startCapture(640, 480, 20)
        println("video stream attributes$videoTrack")
        localVideoTrack!!.setEnabled(true)
        localVideoTrack!!.addSink(localRenderer)
    }
}

private operator fun <K, V> HashMap<K, V>.set(streamId: K?, value: V?) {

}
